src/Simplify_Cast.cpp

```diff
@@ -4,10 +4,16 @@ namespace Halide {
 namespace Internal {
 
 Expr Simplify::visit(const Cast *op, ExprInfo *bounds) {
-    // We generally don't track bounds through casts, with the
-    // exception of casts that constant-fold to a signed integer, so
-    // we don't need the bounds of the value.
-    Expr value = mutate(op->value, nullptr);
+    Expr value = mutate(op->value, bounds);
+
+    if (bounds) {
+        if (bounds->min_defined && !op->type.can_represent(bounds->min)) {
```

Comment:
if the type can't represent the min, then both the min and the max should be undefined, not just the min